### ISSUE TITLE
Reverse-searching (^R) a long command crashes erl on Alpine Linux

### ISSUE THREAD

**Issue by dorian-marchal** [erlang/otp#5116]
**Describe the bug**

In `erl` (or `iex`), when using the reverse-search (Ctrl+R) to get a long command (about 50 chars), the REPL crashes.

**To Reproduce**
```
docker run --rm -it alpine

apk add erlang

erl

1> "a short string"

*do Ctrl+R here, then press Enter: no problem*

1> "this is a very long string that will crash erl"

*do Ctrl+R here, then press Enter: erl crashes with no error message*
```

Example:
<details>

```
$ docker run --rm -it alpine
/ # apk add erlang
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/5) Installing libgcc (10.3.1_git20210424-r2)
(2/5) Installing ncurses-terminfo-base (6.2_p20210612-r0)
(3/5) Installing ncurses-libs (6.2_p20210612-r0)
(4/5) Installing libstdc++ (10.3.1_git20210424-r2)
(5/5) Installing erlang (24.0-r0)
Executing busybox-1.33.1-r3.trigger
OK: 71 MiB in 19 packages
/ # erl
Erlang/OTP 24 [erts-12.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:no-native-stack]

Eshell V12.0  (abort with ^G)
1> "a short string"
(search)`': "a short string"
1> "this is a very long string that will crash erl"
(search)`Illegal instructionlong string that will crash erl"
                            / #
```
</details>

**Affected versions**

```
# erl --version
Erlang/OTP 24 [erts-12.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:no-native-stack]
```

**Additional context**

Original issue: https://github.com/hexpm/bob/issues/96

**Comment by garazdawi** [comment]
This works fine for me. What OS are you running docker in?

**Comment by dorian-marchal** [comment]
We've reproduced the issue on these machines in our team:

![Screenshot 2021-08-13 at 09 43 42](https://user-images.githubusercontent.com/6225979/129322498-2deebe50-3339-40c0-a531-a87f27ccf2a4.png)

<img width="330" alt="Capture_decran_2021-08-13_a_09 45 09" src="https://user-images.githubusercontent.com/6225979/129323453-bb42957c-6247-4ee4-b18f-62bcc87db22f.png">

And a Linux one, too (but I don't have the specs, yet, I'll add them here later).

To be sure, once you pasted the long string (this one: `"this is a very long string that will crash erl"`) did you hit <kbd>Ctrl + R</kbd>, *then* <kbd>Enter</kbd>?

Thanks!

**Comment by garazdawi** [comment]
I'm still unable to reproduce it. Is this running alpine in docker and installing erlang via apk on all of these machines?

**Comment by dorian-marchal** [comment]
Yes, exactly the same steps as described.
In fact, we could not find a machine on which these steps doesn't crash when running (apk-installed) erl from the `alpine` image:

1.  ```
    erl
    ```

2.  ```
    1> "this is a very long string that will crash erl"
    ```

3. <kbd>Enter</kbd>

4. <kbd>Ctrl + R</kbd>

5. <kbd>Enter</kbd>


Here is a recording:


https://user-images.githubusercontent.com/6225979/129392880-2905e1d4-76f4-41b8-861e-8a017aecf9f0.mov

**Comment by mikpe** [comment]
I can't reproduce either, but I'm on a glibc system while Alpine apparently uses musl. In the video I see that the crash is a SIGILL, it would be useful if you could attach gdb to beam.smp before running the crash recipe, and capture what gdb says about the crash, including a stack dump, register dump, and disassembly of the nearby code.

I also see that your erl's banner says "[jit:no-native-stack]" while mine just says "[jit]".

**Comment by dorian-marchal** [comment]
Here is the gdb log:

```
(gdb) cont
Continuing.

Thread 7 "3_scheduler" received signal SIGILL, Illegal instruction.
[Switching to LWP 28]
0x000055c31913b59d in ?? ()
(gdb)
Continuing.
Couldn't get registers: No such process.
Couldn't get registers: No such process.
(gdb)
Continuing.
Couldn't get registers: No such process.
Couldn't get registers: No such process.
(gdb)
Continuing.
Couldn't get registers: No such process.
Couldn't get registers: No such process.
(gdb) [LWP 53 exited]
[LWP 52 exited]
[LWP 51 exited]
[LWP 50 exited]
[LWP 49 exited]
[LWP 48 exited]
[LWP 47 exited]
[LWP 46 exited]
[LWP 45 exited]
[LWP 44 exited]
[LWP 43 exited]
[LWP 42 exited]
[LWP 41 exited]
[LWP 40 exited]
[LWP 39 exited]
[LWP 38 exited]
[LWP 37 exited]
[LWP 36 exited]
[LWP 35 exited]
[LWP 34 exited]
[LWP 33 exited]
[LWP 32 exited]
[LWP 31 exited]
[LWP 30 exited]
[LWP 29 exited]
[LWP 27 exited]
[LWP 26 exited]
[LWP 24 exited]
[LWP 23 exited]
[LWP 22 exited]
[LWP 18 exited]

Program terminated with signal SIGILL, Illegal instruction.
The program no longer exists.
```

(I'm not used to this, I started the container with `docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -it alpine sh`, ran `apk add erlang gdb` and attached to `beam.smp` with `gdb attach -p <pid>`)

**Comment by garazdawi** [comment]
I managed to reproduce it. I was missing one of the "enter"s needed to reproduce.

Now I just have to figure out what is going on.

**Comment by garazdawi** [comment]
Found the problem. The reason why it only shows up in alpine is that glibc seems to not check buffer overlaps as aggressive as musl.

**Comment by dorian-marchal** [comment]
Hi @garazdawi, thanks for the quick fix. 🚀 
Do you know if it will be released in an OTP 24.0.* version?

**Comment by garazdawi** [comment]
Yes, it will be released in the next 22, 23 and 24 patches.
